File: 724.find-pivot-index.java
/*
 * @lc app=leetcode id=724 lang=java
 *
 * [724] Find Pivot Index
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public int pivotIndex(int[] nums) {
        if (nums.length == 1) {
            return 0;
        }
        int sum = Arrays.stream(nums).sum();
        int acc = 0;
        int ix = 0;
        while (ix != nums.length) {
            if (acc * 2 + nums[ix] == sum) break;
            acc += nums[ix++];
        }
        return (ix == nums.length) ? -1 : ix;
    }
}
// @lc code=end

File: 724.find-pivot-index.cpp
/*
 * @lc app=leetcode id=724 lang=cpp
 *
 * [724] Find Pivot Index
 */

// @lc code=start

#include <vector>
#include <numeric>

using namespace std;

class Solution {
public:
    int pivotIndex(vector<int>& nums) {
        if (nums.size() == 1) return 0;
        vector<int> prefix_sum(nums.size()), suffix_sum(nums.size());
        partial_sum(nums.begin(), nums.end(), prefix_sum.begin());
        partial_sum(nums.rbegin(), nums.rend(), suffix_sum.rbegin());
        for (unsigned ix = 0; ix != nums.size(); ++ix) {
            if (prefix_sum[ix] == suffix_sum[ix]) return ix;
        }
        return -1;
    }
};
// @lc code=end
